### PR TITLE
only allow deployments of verified users

### PR DESCRIPTION
**Pull request by Eslam-Nawara**
### Description

only allow deployments of verified users

### Related Issues

- #1234 

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

### CODE REVIEW

**Comment by Eslam-Nawara** [comment]
https://github.com/threefoldtech/zos/pull/2445#issuecomment-2435204123

File: grid-client/deployer/tf_plugin_client.go [hunk 1]
```diff
@@ -346,3 +353,45 @@ func (t *TFPluginClient) BatchCancelContract(contracts []uint64) error {
 func generateSessionID() string {
 	return fmt.Sprintf("tf-%d", os.Getpid())
 }
+
+// make sure the account used is verified we have the user public key in bytes(pkBytes)
+func getTwinVerificationState(twinID uint32) (status string) {
+	verificationServiceURL := "https://kyc1.gent01.dev.grid.tf/api/v1/status"
```
Comment (xmonader) [inline grid-client/deployer/tf_plugin_client.go]: this url should be retrieved from stack information

File: grid-client/deployer/tf_plugin_client.go [hunk 2]
```diff
@@ -271,6 +275,18 @@ func NewTFPluginClient(
 	}
 	tfPluginClient.TwinID = twinID
 
+	// make sure the account used is verified
+	check := func() error {
+		if !isTwinVerified(twinID, tfPluginClient.Network) {
+			return fmt.Errorf("user is not verified")
```
Comment (xmonader) [inline grid-client/deployer/tf_plugin_client.go]: encode twin id in the error please

File: grid-client/deployer/tf_plugin_client.go [hunk 3]
```diff
@@ -271,6 +275,18 @@ func NewTFPluginClient(
...
+		}
+		return nil
+	}
+
+	if err := backoff.Retry(check, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 5)); err != nil {
+		return TFPluginClient{}, fmt.Errorf("can not run deployments for unverified user, please visit https://dashboard.grid.tf/ to verify your account")
```
Comment (xmonader) [inline grid-client/deployer/tf_plugin_client.go]: only verified users can deploy, please .....

File: grid-client/deployer/tf_plugin_client.go [hunk 4]
```diff
@@ -346,3 +362,46 @@ func (t *TFPluginClient) BatchCancelContract(contracts []uint64) error {
 func generateSessionID() string {
 	return fmt.Sprintf("tf-%d", os.Getpid())
 }
+
+// isTwinVerified makes sure the twin used is verified
+func isTwinVerified(twinID uint32, net string) (verified bool) {
```
Comment (rawdaGastan) [inline grid-client/deployer/tf_plugin_client.go]: why not returning the errors? then we can log them if not needed when using the function

**Review by rawdaGastan — APPROVED** [review]

File: grid-client/deployer/stacks.go [hunk 5]
```diff
@@ -93,4 +93,11 @@ var (
 			"wss://relay.02.grid.tf",
 		},
 	}
+
+	KycURLs = map[string]string{
+		DevNetwork:  "https://kyc.dev.grid.tf ",
```
Comment (rawdaGastan) [inline grid-client/deployer/stacks.go]: remove extra space :D

**Comment by Eslam-Nawara** [comment]
workflows are failing due to using unverified account `Alice` reported here https://github.com/threefoldtech/tfgrid-sdk-go/issues/1239

**Pull request merged**